File: src/main/scala/keri/projectx/client/render/RenderSimpleGlow.java
/*
 * Copyright (c) 2017 KitsuneAlex & Adam8234. All rights reserved!
 * Do not destribute or redistribute this software without the
 * explicit permission of the developer!
 */

package keri.projectx.client.render;

import codechicken.lib.render.CCModel;
import codechicken.lib.render.CCRenderState;
import codechicken.lib.render.buffer.BakingVertexBuffer;
import codechicken.lib.vec.Cuboid6;
import codechicken.lib.vec.Translation;
import codechicken.lib.vec.Vector3;
import codechicken.lib.vec.uv.IconTransformation;
import com.google.common.collect.Lists;
import keri.ninetaillib.lib.render.IBlockRenderingHandler;
import keri.ninetaillib.lib.render.RenderingRegistry;
import keri.ninetaillib.lib.texture.IIconBlock;
import keri.ninetaillib.lib.util.BlockAccessUtils;
import keri.ninetaillib.lib.util.RenderUtils;
import keri.projectx.client.render.connected.BlockRenderContext;
import keri.projectx.client.render.connected.ICTMBlock;
import net.minecraft.block.Block;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.OpenGlHelper;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.VertexBuffer;
import net.minecraft.client.renderer.block.model.BakedQuad;
import net.minecraft.client.renderer.texture.TextureAtlasSprite;
import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
import net.minecraft.item.ItemStack;
import net.minecraft.util.BlockRenderLayer;
import net.minecraft.util.EnumBlockRenderType;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.IBlockAccess;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import org.lwjgl.opengl.GL11;

import java.util.List;

@SideOnly(Side.CLIENT)
public class RenderSimpleGlow implements IBlockRenderingHandler {

    public static EnumBlockRenderType RENDER_TYPE;
    private static CCModel BLOCK_MODEL;

    static{
        RENDER_TYPE = RenderingRegistry.getNextAvailableType();
        RenderingRegistry.registerRenderingHandler(new RenderSimpleGlow());
        BLOCK_MODEL = CCModel.quadModel(24).generateBlock(0, new Cuboid6(0D, 0D, 0D, 1D, 1D, 1D)).computeNormals();
    }

    @Override
    public boolean renderWorld(IBlockAccess world, BlockPos pos, VertexBuffer buffer, BlockRenderLayer layer){
        CCRenderState renderState = CCRenderState.instance();
        IIconBlock iconProvider = (IIconBlock)world.getBlockState(pos).getBlock();
        IAnimationHandler animationHandler = (IAnimationHandler)world.getBlockState(pos).getBlock();
        CCModel modelAnimation = BLOCK_MODEL.copy();
        modelAnimation.apply(new Translation(Vector3.fromBlockPos(pos)));
        renderState.bind(buffer);

        for(int side = 0; side < 6; side++){
            TextureAtlasSprite texture = animationHandler.getAnimationIcon(world, pos, side) ;
            int color = animationHandler.getAnimationColor(world, pos, side);
            int brightness = animationHandler.getAnimationBrightness(world, pos, side);
            renderState.brightness = brightness;
            modelAnimation.setColour(color);
            modelAnimation.render(renderState, 4 * side, 4 + (4 * side), new IconTransformation(texture));
        }

        List<BakedQuad> overlayQuads = Lists.newArrayList();

        if(layer == BlockRenderLayer.CUTOUT_MIPPED){
            if(world.getBlockState(pos).getBlock() instanceof ICTMBlock){
                TextureAtlasSprite texture = null;

                if(iconProvider.getIcon(world, pos, 0) != null){
                    texture = iconProvider.getIcon(world, pos, 0);
                }
                else{
                    texture = iconProvider.getIcon(BlockAccessUtils.getBlockMetadata(world, pos), 0);
                }

                BlockRenderContext renderContext = new BlockRenderContext();
                renderContext.setBlockAccess(world);
                renderContext.setCurrentBlockState(world.getBlockState(pos));
                renderContext.setChangeBounds(true);
                renderContext.renderStandardBlock(pos, texture);
                BakingVertexBuffer parent = BakingVertexBuffer.create();
                parent.begin(GL11.GL_QUADS, DefaultVertexFormats.ITEM);
                renderState.reset();
                renderState.bind(parent);
                renderContext.getModel().render(renderState);
                parent.finishDrawing();
                overlayQuads.addAll(parent.bake());
            }
            else{
                CCModel modelOverlay = BLOCK_MODEL.copy();
                BakingVertexBuffer parent = BakingVertexBuffer.create();
                parent.begin(GL11.GL_QUADS, DefaultVertexFormats.ITEM);
                renderState.reset();
                renderState.bind(parent);

                for(int side = 0; side < 6; side++){
                    TextureAtlasSprite texture = null;

                    if(iconProvider.getIcon(world, pos, side) != null){
                        texture = iconProvider.getIcon(world, pos, side);
                    }
                    else{
                        texture = iconProvider.getIcon(BlockAccessUtils.getBlockMetadata(world, pos), side);
                    }

                    int colorMultiplier = iconProvider.getColorMultiplier(BlockAccessUtils.getBlockMetadata(world, pos), side);
                    modelOverlay.setColour(colorMultiplier);
                    modelOverlay.render(renderState, 4 * side, 4 + (4 * side), new IconTransformation(texture));
                }

                parent.finishDrawing();
                overlayQuads.addAll(parent.bake());
            }
        }

        return RenderUtils.renderQuads(buffer, world, pos, overlayQuads);
    }

    @Override
    public void renderDamage(IBlockAccess world, BlockPos pos, VertexBuffer buffer, TextureAtlasSprite texture) {
        CCModel model = BLOCK_MODEL.copy();
        model.apply(new Translation(Vector3.fromBlockPos(pos)));
        CCRenderState renderState = CCRenderState.instance();
        renderState.bind(buffer);
        model.render(renderState, new IconTransformation(texture));
    }

    @Override
    public void renderInventory(ItemStack stack, VertexBuffer buffer) {
        int lastBrightness = (int)OpenGlHelper.lastBrightnessY << 16 | (int)OpenGlHelper.lastBrightnessX;
        Tessellator.getInstance().draw();
        GlStateManager.pushMatrix();
        GlStateManager.disableLighting();
        RenderUtils.MipmapFilterData mipmapFilterData = RenderUtils.disableMipmap();
        buffer.begin(GL11.GL_QUADS, RenderUtils.getFormatWithLightMap(DefaultVertexFormats.ITEM));
        CCRenderState renderState = CCRenderState.instance();
        IIconBlock iconProvider = (IIconBlock)Block.getBlockFromItem(stack.getItem());
        IAnimationHandler animationHandler = (IAnimationHandler)Block.getBlockFromItem(stack.getItem());
        renderState.bind(buffer);
        CCModel modelAnimation = BLOCK_MODEL.copy();

        for(int side = 0; side < 6; side++){
            TextureAtlasSprite texture = animationHandler.getAnimationIcon(stack, side);
            int color = animationHandler.getAnimationColor(stack, side);
            int brightness = animationHandler.getAnimationBrightness(stack, side);
            renderState.brightness = brightness;
            modelAnimation.setColour(color);
            modelAnimation.render(renderState, 4 * side, 4 + (4 * side), new IconTransformation(texture));
        }

        Tessellator.getInstance().draw();
        RenderUtils.enableMipmap(mipmapFilterData);
        GlStateManager.enableLighting();
        GlStateManager.popMatrix();
        buffer.begin(GL11.GL_QUADS, DefaultVertexFormats.ITEM);
        renderState.reset();
        renderState.bind(buffer);
        renderState.brightness = lastBrightness;
        CCModel modelOverlay = BLOCK_MODEL.copy();

        for(int side = 0; side < 6; side++){
            TextureAtlasSprite texture = iconProvider.getIcon(stack.getMetadata(), side);
            int colorMultiplier = iconProvider.getColorMultiplier(stack.getMetadata(), side);
            modelOverlay.setColour(colorMultiplier);
            modelOverlay.render(renderState, 4 * side, 4 + (4 * side), new IconTransformation(texture));
        }

        Tessellator.getInstance().draw();
        buffer.begin(GL11.GL_QUADS, RenderUtils.getFormatWithLightMap(DefaultVertexFormats.ITEM));
    }

    @Override
    public EnumBlockRenderType getRenderType() {
        return RENDER_TYPE;
    }

}

File: src/main/scala/keri/projectx/client/render/RenderSimpleGlow.scala
/*
 * Copyright (c) 2017 KitsuneAlex & Adam8234. All rights reserved!
 * Do not distribute or redistribute without the explicit permission
 * of the developer!
 */

package keri.projectx.client.render

import codechicken.lib.render.CCRenderState
import codechicken.lib.render.buffer.BakingVertexBuffer
import codechicken.lib.vec.uv.IconTransformation
import codechicken.lib.vec.{Cuboid6, Translation, Vector3}
import keri.ninetaillib.lib.render.connected.{ConnectedRenderContext, ICTMBlock}
import keri.ninetaillib.lib.render.{IBlockRenderingHandler, RenderingRegistry}
import keri.ninetaillib.lib.texture.IIconBlock
import keri.ninetaillib.lib.util.{ModelUtils, RenderUtils}
import net.minecraft.block.Block
import net.minecraft.block.state.IBlockState
import net.minecraft.client.renderer.texture.TextureAtlasSprite
import net.minecraft.client.renderer.vertex.DefaultVertexFormats
import net.minecraft.client.renderer.{GlStateManager, OpenGlHelper, Tessellator, VertexBuffer}
import net.minecraft.item.ItemStack
import net.minecraft.util.math.BlockPos
import net.minecraft.util.{BlockRenderLayer, EnumBlockRenderType, EnumFacing}
import net.minecraft.world.IBlockAccess
import net.minecraftforge.fml.relauncher.{Side, SideOnly}
import org.lwjgl.opengl.GL11

@SideOnly(Side.CLIENT)
object RenderSimpleGlow extends IBlockRenderingHandler {

  private final val BLOCK_MODEL = ModelUtils.getNormalized(new Cuboid6(0D, 0D, 0D, 16D, 16D, 16D))
  final val RENDER_TYPE: EnumBlockRenderType = RenderingRegistry.getNextAvailableType
  RenderingRegistry.registerRenderingHandler(RenderSimpleGlow.this)

  override def renderWorld(world: IBlockAccess, pos: BlockPos, state: IBlockState, buffer: VertexBuffer, layer: BlockRenderLayer): Boolean = {
    val iconProvider = state.getBlock.asInstanceOf[IIconBlock]
    val animationHandler = state.getBlock.asInstanceOf[IAnimationHandler]
    val renderState = CCRenderState.instance
    val translation = new Translation(Vector3.fromBlockPos(pos))
    val meta = state.getBlock.getMetaFromState(state)
    renderState.reset()
    renderState.bind(buffer)

    for (side <- 0 until 6) {
      val texture = animationHandler.getAnimationIcon(world, pos, side)
      val color = animationHandler.getAnimationColor(world, pos, side)
      renderState.brightness = animationHandler.getAnimationBrightness(world, pos, side)
      BLOCK_MODEL.copy.apply(translation).setColour(color).render(renderState, 4 * side, 4 + 4 * side, new IconTransformation(texture))
    }

    if (layer == BlockRenderLayer.CUTOUT_MIPPED) {
      val parentBuffer = BakingVertexBuffer.create
      parentBuffer.begin(GL11.GL_QUADS, DefaultVertexFormats.ITEM)
      renderState.reset
      renderState.bind(parentBuffer)

      state.getBlock match {
        case ctmBlock: ICTMBlock => {
          val renderContext: ConnectedRenderContext = new ConnectedRenderContext

          for (side <- 0 until 6) {
            val texture: Array[TextureAtlasSprite] = ctmBlock.getConnectedTexture(world, pos, EnumFacing.getFront(side))
            renderContext.setBlockAccess(world)
            renderContext.setCurrentBlockState(state)
            renderContext.setChangeBounds(true)

            if (ctmBlock.canTextureConnect(world, pos, EnumFacing.getFront(side))) {
              renderContext.renderFace(pos, texture, EnumFacing.getFront(side))
            }
          }

          renderContext.getModel.copy.render(renderState)
        }
        case _ => {
          for (side <- 0 until 6) {
            var texture: TextureAtlasSprite = null
            val color = iconProvider.getColorMultiplier(meta, EnumFacing.getFront(side))

            if (iconProvider.getIcon(world, pos, EnumFacing.getFront(side)) != null) {
              texture = iconProvider.getIcon(world, pos, EnumFacing.getFront(side))
            }
            else {
              texture = iconProvider.getIcon(meta, EnumFacing.getFront(side))
            }

            BLOCK_MODEL.copy.setColour(color).render(renderState, 4 * side, 4 + 4 * side, new IconTransformation(texture))
          }
        }
      }

      parentBuffer.finishDrawing()
      RenderUtils.renderQuads(buffer, world, pos, parentBuffer.bake())
    }

    false
  }

  override def renderDamage(world: IBlockAccess, pos: BlockPos, state: IBlockState, buffer: VertexBuffer, texture: TextureAtlasSprite): Unit = {
    val renderState = CCRenderState.instance
    val translation = new Translation(Vector3.fromBlockPos(pos))
    renderState.reset
    renderState.bind(buffer)
    BLOCK_MODEL.copy.apply(translation).render(renderState, new IconTransformation(texture))
  }

  override def renderInventory(stack: ItemStack, buffer: VertexBuffer): Unit = {
    val iconProvider = Block.getBlockFromItem(stack.getItem).asInstanceOf[IIconBlock]
    val animationHandler = Block.getBlockFromItem(stack.getItem).asInstanceOf[IAnimationHandler]
    val meta = stack.getMetadata
    val lastBrightness = OpenGlHelper.lastBrightnessY.toInt << 16 | OpenGlHelper.lastBrightnessX.toInt
    val renderState = CCRenderState.instance
    Tessellator.getInstance.draw
    GlStateManager.pushMatrix
    GlStateManager.disableLighting
    val filterData = RenderUtils.disableMipmap
    buffer.begin(GL11.GL_QUADS, RenderUtils.getFormatWithLightMap(DefaultVertexFormats.ITEM))
    renderState.reset
    renderState.bind(buffer)

    for (side <- 0 until 6) {
      val texture = animationHandler.getAnimationIcon(stack, side)
      val color = animationHandler.getAnimationColor(stack, side)
      renderState.brightness = animationHandler.getAnimationBrightness(stack, side)
      BLOCK_MODEL.copy.setColour(color).render(renderState, 4 * side, 4 + 4 * side, new IconTransformation(texture))
    }

    Tessellator.getInstance.draw
    GlStateManager.popMatrix
    GlStateManager.pushMatrix()
    GlStateManager.enableLighting
    RenderUtils.enableMipmap(filterData)
    buffer.begin(GL11.GL_QUADS, DefaultVertexFormats.ITEM)
    renderState.reset
    renderState.bind(buffer)
    renderState.brightness = lastBrightness

    for (side <- 0 until 6) {
      val texture = iconProvider.getIcon(meta, EnumFacing.getFront(side))
      val color = iconProvider.getColorMultiplier(meta, EnumFacing.getFront(side))
      BLOCK_MODEL.copy.setColour(color).render(renderState, 4 * side, 4 + 4 * side, new IconTransformation(texture))
    }

    Tessellator.getInstance.draw
    GlStateManager.popMatrix
    buffer.begin(GL11.GL_QUADS, RenderUtils.getFormatWithLightMap(DefaultVertexFormats.ITEM))
  }

  override def getRenderType: EnumBlockRenderType = RENDER_TYPE

}
